### PR TITLE
[REVIEW]add GetIncrementalUsers func

### PR DESCRIPTION
**Pull request by XiaokaiPhil**
GetIncrementalUsers will return a list of users updated from a specific time point. ([Asana link](https://app.asana.com/0/1121483390875723/1136442831379687?lg=1565647977479))

**NOTE**: in the current stage, some variables are hardcoded which should be removed in the future optimization steps.

### CODE REVIEW

File: zendesk/zd_user_service.go [hunk 1]
```diff
@@ -161,6 +161,97 @@ func (c *client) AddUserTags(id int64, tags []string) ([]string, error) {
...
+
+	headers := map[string]string{}
+	if in != nil {
+		headers["Content-Type"] = "application/json"
+	}
+
+	url := "https://philhelp.zendesk.com/api/v2/incremental/users.json?start_time="
+	apiV2 := "/api/v2/incremental/users.json?start_time="
+	apiStartIndex := strings.Index(url, apiV2)
+	endpoint := fmt.Sprintf("%s%v", "/api/v2/incremental/users.json?start_time=", unixTime)
```
Comment (abhyuday) [inline zendesk/zd_user_service.go]: use apiV2. You can use it in url and endpoint

File: zendesk/zd_user_service.go [hunk 2]
```diff
@@ -161,6 +161,97 @@ func (c *client) AddUserTags(id int64, tags []string) ([]string, error) {
...
+	res, err := c.request("GET", endpoint, headers, bytes.NewReader(payload))
+	defer res.Body.Close()
+
+	dataPerPage := new(APIPayload)
```
Comment (abhyuday) [inline zendesk/zd_user_service.go]: initialize variable where needed.

File: zendesk/zendesk_client_service.go [hunk 3]
```diff
@@ -65,6 +65,7 @@ type Client interface {
 	GetAllTicketMetrics() ([]TicketMetric, error)
 	ShowTicketMetric(int64) (*TicketMetric, error)
 	GetAllTicketComments() (map[int][]TicketComment, error)
+	GetIncrementalUsers(int64) ([]User, error)
```
Comment (abhyuday) [inline zendesk/zendesk_client_service.go]: Call this GetUsersIncrementally() as that makes more sense.

File: zendesk/zd_user_service.go [hunk 4]
```diff
@@ -161,6 +161,97 @@ func (c *client) AddUserTags(id int64, tags []string) ([]string, error) {
...
+	return getUniqUsers(result), err
+}
+
+// getUniqUsers is to remove the duplicate records due to pagination
+// more details can be found int the following link
+// https://developer.zendesk.com/rest_api/docs/support/incremental_export#excluding_pagination_duplicates
+
+func getUniqUsers(users []User) []User {
+	var Empty struct{}
+	keys := make(map[string]struct{})
```
Comment (abhyuday) [inline zendesk/zd_user_service.go]: you can use a map of 
m := map[string]bool{} where if m[key] exists, then m[key] == true. No need for empty structs and all

Comment (XiaokaiPhil) [reply zendesk/zd_user_service.go]: https://yourbasic.org/golang/implement-set/
this post explains two ways: one uses boolean as you mentioned which is more intuitive. the other one uses empty struct, the advantage of which is that it takes no memory

**Pull request merged**